src/HexagonOptimize.cpp

```diff
@@ -0,0 +1,1117 @@
+                   yields_removable_interleave(args)) {
+            // This call has a deinterleaving alternative, and the
+            // arguments are interleaved, so we should use the
+            // alternative instead.
+            const DeinterleavingAlternative &alt = deinterleaving_alts[op->name];
+            for (Expr &i : args) {
+                i = remove_interleave(i);
+            }
+            for (Expr i : alt.extra_args) {
+                args.push_back(i);
+            }
+            expr = Call::make(op->type, alt.name, args, op->call_type);
+        } else if (changed) {
+            expr = Call::make(op->type, op->name, args, op->call_type,
+                              op->func, op->value_index, op->image, op->param);
+        } else {
+            expr = op;
+        }
+    }
+
+    using IRMutator::visit;
+};
+
+// Find a conservative upper bound of an expression.
+class UpperBound : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const Sub *op) {
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        const Min *min_a = a.as<Min>();
+        const Min *min_b = b.as<Min>();
+        const Max *max_a = a.as<Max>();
+        const Max *max_b = b.as<Max>();
+
+        if (max_a && max_b) {
+            if (equal(max_a->b, max_b->b)) {
+                expr = mutate(simplify(max_a->a - max_b->a));
+                return;
+            }
+        }
+
+        if (min_a && min_b) {
+            if (equal(min_a->b, min_b->b)) {
+                expr = mutate(simplify(min_a->a - min_b->a));
+                return;
+            }
+        }
+
+        if (!a.same_as(op->a) || !b.same_as(op->b)) {
+            expr = Sub::make(a, b);
+        } else {
+            expr = op;
+        }
+    }
+};
+
+Expr upper_bound(Expr x) {
+    return simplify(UpperBound().mutate(x));
+}
+
+// Replace indirect loads with dynamic_shuffle intrinsics where
+// possible.
+class OptimizeShuffles : public IRMutator {
+    Scope<Interval> bounds;
+
+    using IRMutator::visit;
+
+    template <typename T>
+    void visit_let(const T *op) {
+        // We only care about vector lets.
+        if (op->value.type().is_vector()) {
+            bounds.push(op->name, bounds_of_expr_in_scope(op->value, bounds));
+        }
+        IRMutator::visit(op);
+        if (op->value.type().is_vector()) {
+            bounds.pop(op->name);
+        }
+    }
+
+    void visit(const Let *op) { visit_let(op); }
+    void visit(const LetStmt *op) { visit_let(op); }
+
+    void visit(const Load *op) {
+        if (!op->type.is_vector() || op->index.as<Ramp>()) {
+            // Don't handle scalar or simple vector loads.
+            IRMutator::visit(op);
+            return;
+        }
+
+        Expr index = mutate(op->index);
+        Interval index_bounds = bounds_of_expr_in_scope(index, bounds);
+        Expr index_span = index_bounds.max - index_bounds.min;
```

Comment:
This line might crash - bounds_of_expr_in_scope can return half-open intervals. (e.g. [x, pos_inf])